crates/abd-clam/src/core/graph/criteria.rs

```diff
@@ -3,25 +3,103 @@
 
+use std::cmp::Ordering;
+use std::collections::{BinaryHeap, HashSet};
+
 use crate::core::graph::_graph::{ClusterSet, EdgeSet};
 use crate::{Cluster, Dataset, Edge, Instance};
 use distances::Number;
-use std::collections::HashSet;
 
-/// Filler function to select clusters for graph
+/// A Wrapper that contains a cluster and its score
+pub struct ClusterWrapper<'a, U: Number> {
+    /// A cluster
+    pub cluster: &'a Cluster<U>,
+    /// An associated score
+    pub score: f64,
+}
+
+impl<'a, U: Number> PartialEq for ClusterWrapper<'a, U> {
+    fn eq(&self, other: &Self) -> bool {
+        self.score == other.score
+    }
+}
+
+impl<'a, U: Number> Eq for ClusterWrapper<'a, U> {}
+
+impl<'a, U: Number> Ord for ClusterWrapper<'a, U> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.score.partial_cmp(&other.score).unwrap_or(Ordering::Equal)
+    }
+}
+
+impl<'a, U: Number> PartialOrd for ClusterWrapper<'a, U> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Scores a cluster with a given scoring fuction
+///
+///
+/// # Arguments
+///
+/// * `root`: The root of the tree.
+/// * `scoring_function`: Function in which to score each cluster
+///
+/// # Returns:
+///
+/// `BinaryHeap` of `ClusterWrappers`
+///
+pub fn score_clusters<'a, U: Number>(
+    root: &'a Cluster<U>,
+    scoring_function: &crate::core::graph::MetaMLScorer,
+) -> Result<BinaryHeap<ClusterWrapper<'a, U>>, String> {
+    let mut scored_clusters: BinaryHeap<ClusterWrapper<'a, U>> = BinaryHeap::new();
+
+    for cluster in root.subtree() {
+        let score = match cluster.ratios() {
+            Some(ratios) => scoring_function(ratios),
+            None => return Err("Error: tree must be built with ratios".to_string()),
+        };
+        scored_clusters.push(ClusterWrapper { cluster, score });
+    }
+
+    Ok(scored_clusters)
+}
+
+/// Gets `ClusterSet` from `BinaryHeap` of `ClusterWrappers`
+///
+/// # Arguments
+///
+/// * clusters : `BinaryHeap` of `ClusterWrappers` containing a cluster and its score
+///
+/// # Returns:
+///
+/// `ClusterSet` of chosen clusters representing highest scored with no ancestors or descendants
+///
+/// # Errors
+///
+/// If `ClusterWrapper` contains an invalid cluster-score pairing
+///
 pub fn select_clusters<'a, U: Number>(
     root: &'a Cluster<U>,
-    _scorer_function: &MetaMLScorer,
-    depth: usize,
-) -> ClusterSet<'a, U> {
-    // TODO! Replace with proper cluster selection algorithm
-    let mut selected_clusters = ClusterSet::new();
-    for c in root.subtree() {
-        if c.depth() == depth || (c.depth() < depth && c.is_leaf()) {
-            selected_clusters.insert(c);
-        }
+    scoring_function: &MetaMLScorer,
+) -> Result<ClusterSet<'a, U>, String> {
+    let mut cluster_set: HashSet<&'a Cluster<U>> = HashSet::new();
+    let mut clusters = score_clusters(root, scoring_function)?;
+
+    clusters.retain(|item| item.cluster.depth() > 3);
```

Comment:
This will remove any leaves at a lower depth. You need to keep such leaves around. Also, the minimum depth should be an input parameter.

Reply:
Fixed